### PR TITLE
ImmutableMatrix creates MatrixObj for Z/nZ

### PR DESCRIPTION
**Pull request by hulpke**
and subsequent changes in group, meataxe, and polynomial code to make it work. This also provides further  tests for #4773

The result is a small speedup when working with matrix groups/modules over ZmodnZ rings.

## Text for release notes

N/A

### CODE REVIEW

File: tst/teststandard/opers/Matobjnz.tst [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+gap> g:=AlternatingGroup(5);;
+gap> mo:=IrreducibleModules(g,GF(p));;
+gap> Set(List(mo[2],x->x.dimension));
+[ 1, 4, 5, 6 ]
+gap> h:=Group(mo[2][2].generators);
+<matrix group with 2 generators>
+gap> Size(h);
+60
+gap> Length(ConjugacyClassesSubgroups(h));
+9
```
Comment (fingolfin) [inline tst/teststandard/opers/Matobjnz.tst:14]: These two tests fail with errors

File: lib/vecmat.gi [hunk 2]
```diff
@@ -1586,10 +1595,16 @@ local sf, rep, ind, ind2, row, i,big,l;
     rep:=IsPlistRep;
   fi;
 
+  if IsPlistRep(matrix) and not ForAll(matrix,IsPlistRep) then
+    nr:=Length(matrix);
+  else
+    nr:=NrRows(matrix);
+  fi;
+
```
Comment (fingolfin) [inline lib/vecmat.gi:1603]: Why is this necessary? `NrRows` should work for "old-style" matrices, too

File: lib/grpmat.gi [hunk 3]
```diff
@@ -972,6 +979,81 @@ local G,fam,typ,f;
...
+local G,typ,f;
+
+  if not IsFinite(gens) then TryNextMethod(); fi;
+  typ:=MakeGroupyType(FamilyObj(gens), IsGroup and IsAttributeStoringRep 
+          and HasGeneratorsOfMagmaWithInverses and IsFinitelyGeneratedGroup
+          and HasIsEmpty and IsFinite and HasOne,
+          gens,id,true);
+
+  f:=DefaultScalarDomainOfMatrixList(gens);
+  gens:=List(Immutable(gens),i->ImmutableMatrix(f,i));
```
[the remark this answers is not in the document]
Comment (hulpke) [reply lib/grpmat.gi:1021]: I suspect this is the result of a global insertion of `ImmutableMatrix` in many places without considering in each place whether the local code could be improved subsequently.

File: lib/vecmat.gi (hunk 2 above)
Comment (hulpke) [reply lib/vecmat.gi]: Did not realize that -- chaned.

File: tst/teststandard/opers/Matobjnz.tst (hunk 1 above)
Comment (hulpke) [reply tst/teststandard/opers/Matobjnz.tst:14]: y worked initially, but subsequent changed had made them fail. Fixed again.

File: lib/vecmat.gi (hunk 2 above)
Comment (hulpke) [reply lib/vecmat.gi]: Actually...
It is needed, because the input could be a mixed format list, for which `NrRows` will not work.

Comment (fingolfin) [reply lib/vecmat.gi]: Could you elaborate on that a bit? What is a "mixed format list" and where does it occur?

**Review by fingolfin — APPROVED** [review]
Let's just merge this

File: lib/vecmat.gi (hunk 2 above)
Comment (hulpke) [reply lib/vecmat.gi]: While merged already, a brief explanation for later:

The function is `ImmutableMatrix`. It takes somethig that could be construed as a list of vectors, and converts it into a compact format matrix (with the implicit claim that this representation will be efficient to work with).
This could be a list of vectors of different storage/mutability/base ring status (and indeed, this can happen when calling from the MeatAxe).
